Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs

```diff
@@ -1928,7 +1928,9 @@ internal TValue ReadValue<TValue>(int bindingIndex, int controlIndex, bool ignor
                     throw new InvalidOperationException(
                         $"Cannot read value of type '{TypeHelpers.GetNiceTypeName(typeof(TValue))}' from control '{control.path}' bound to action '{GetActionOrNull(bindingIndex)}' (control is a '{control.GetType().Name}' with value type '{TypeHelpers.GetNiceTypeName(control.valueType)}')");
 
-                value = controlOfType.ReadValue();
+                // Avoid reading from controls of devices which are being destroyed.
+                if (control.device.m_DeviceIndex != InputDevice.kInvalidDeviceIndex)
```

Comment:
On the fence about this one. Strictly speaking, this fixes a symptom. Conceptually, what should happen is that all actions referring to the device and that are enabled, should get disabled. Then the device should be removed. Then all actions that were disabled, should get reenabled. In-between these, actions that simply reference controls on the device but aren't enabled should simply have their states wiped.

The way the code does it instead ATM is running it all in one step when the setup is already invalid. Strictly speaking, `InputAction.controls` should never contain controls that aren't added to the system.

My guess is that making it behave the way it should isn't a trivial fix, though, given the way `LazyResolveBindings` operates. Gut feeling is that, given how organically some of that stuff has grown over the many iterations of the action system, there's likely a much simpler way to do all this which cuts a lot of crap from the current logic.

Reply:
@Rene-Damm hmm, yeah. 

So how about this approach (tested locally, works)?

1. Add an `InputDevice.removing` flag
2. Making bindings resolving ignore devices with `removing` flag
3. In RemoveDevice, after calling `RemoveStateChangeMonitors`, set `device.removing = true;`, and zero the device state.
4. When done removing set `device.removing = false;`, in case the device gets reused.

Reply:
Internal `InputDevice.removing`, sounds reasonable to me.

We'd also be moving the `InputActionState.OnDeviceState` call almost all the way up in `InputManager.RemoveDevice`, right?

Reply:
Pushed the change.